experimental/type-checker/src/KindChecker.hs

```diff
@@ -0,0 +1,339 @@
+    KVar a -> pretty a
+
+data Type
+  = Var Atom
+  | App Type Type
+  | Abs String Type
+  deriving (Eq, Show)
+
+instance Pretty Type where
+  pretty = \case
+    Var a -> pretty a
+    App t1 t2 -> show' t1 <> pretty " " <> show' t2
+    Abs a t1 -> pretty "λ" <> pretty a <> pretty "." <> pretty t1
+    where
+      show' = \case
+        Var a -> pretty a
+        App t1 t2 -> parens $ show' t1 <+> show' t2
+        Abs a t1 -> parens $ pretty "λ" <> pretty a <> pretty "." <> show' t1
+
+data Context = Context
+  { context :: [(Atom, Kind)]
+  , addContext :: [(Atom, Kind)]
+  }
+  deriving (Show, Eq)
+
+instance Pretty Context where
+  pretty c = case addContext c of
+    [] -> pretty "Γ"
+    ctx -> pretty "Γ" <+> pretty "∪" <+> braces (setPretty ctx)
+    where
+      setPretty = hsep . punctuate comma . fmap (\(v, t) -> pretty v <> pretty ":" <+> pretty t)
+
+-- | Utility to unify the two.
+getAllContext :: Context -> [(Atom, Kind)]
+getAllContext c = context c <> addContext c
+
+newtype Judgement = Judgement {getJudgement :: (Context, Type, Kind)}
+  deriving (Show, Eq)
+
+instance Pretty Judgement where
+  pretty (Judgement (c, t, k)) = pretty c <> pretty " ⊢ " <> pretty t <+> pretty ":" <+> pretty k
+
+data Derivation
+  = Axiom Judgement
+  | Abstraction Judgement Derivation
+  | Application Judgement Derivation Derivation
+  deriving (Show, Eq)
+
+instance Pretty Derivation where
+  pretty x = case x of
+    Axiom j -> hang 2 $ pretty j
+    Abstraction j d -> dNest j [d]
+    Application j d1 d2 -> dNest j [d1, d2]
+    where
+      dNest j ds = pretty j <> line <> hang 2 (encloseSep (lbracket <> space) rbracket (space <> pretty "∧" <> space) (pretty <$> ds))
+
+type DError = String
+
+newtype Constraint = Constraint (Kind, Kind)
+  deriving (Show, Eq)
+
+instance Pretty Constraint where
+  pretty (Constraint (t1, t2)) = pretty t1 <+> pretty "=" <+> pretty t2
+
+newtype Substitution = Substitution {getSubstitution :: (Atom, Kind)}
+  deriving (Show, Eq)
+
+instance Pretty Substitution where
+  pretty (Substitution (a, k)) = pretty a <+> pretty "↦" <+> pretty k
+
+newtype DerivationContext = DC
+  { freshVarStream :: [Atom]
+  }
+
+type DeriveM a = ReaderT Context (StateT DerivationContext (WriterT [Constraint] (Except DError))) a
+
+type Derive a =
+  forall m.
+  ( MonadReader Context m
+  , MonadState DerivationContext m
+  , MonadWriter [Constraint] m
+  , MonadError DError m
+  ) =>
+  m a
+
+-- | Run derivation builder - not unified yet.
+runDerive :: Type -> Either DError (Derivation, [Constraint])
+runDerive t = runExcept $ runWriterT $ evalStateT (runReaderT (derive t) defContext) (DC atoms)
+
+-- | Run derivation - throw error.
+runDerive' :: Type -> (Derivation, [Constraint])
+runDerive' t = either error id $ runExcept $ runWriterT $ evalStateT (runReaderT (derive t) defContext) (DC atoms)
+
+-- | Creates the derivation
+derive :: Type -> Derive Derivation
+derive x = do
+  c <- ask
```

Comment:
Ok, to understand this I needed to re-trace your steps, I am very slow unfortunately and that's the only way I can provide meaningful feedback.

https://github.com/mlabs-haskell/lambda-buffers/pull/9 is the work of doing the same thing you did with fd-unification library